packages/parcel-optimizer/src/index.ts

```diff
@@ -1,8 +1,108 @@
+import assert from 'assert';
+import { join, basename } from 'path';
+
 import { sort } from '@compiled/css';
+import { hash } from '@compiled/utils';
 import { Optimizer } from '@parcel/plugin';
+import posthtml from 'posthtml';
+import { insertAt } from 'posthtml-insert-at';
+
+import type { ParcelOptimizerOpts } from './types';
+
+const configFiles = [
+  '.compiledcssrc',
+  '.compiledcssrc.json',
+  'compiledcss.js',
+  'compiledcss.config.js',
+];
+
+export default new Optimizer<ParcelOptimizerOpts>({
+  async loadConfig({ config, options }) {
+    const conf = await config.getConfigFrom(join(options.projectRoot, 'index'), configFiles, {
+      packageKey: '@compiled/parcel-optimizer',
+    });
+
+    const contents = {
+      inlineCss: false,
+    };
+
+    if (conf) {
+      config.invalidateOnStartup();
+      Object.assign(contents, conf.contents);
+    }
+
+    return contents;
+  },
+
+  async optimize({ contents, map, bundle, bundleGraph, options, config }) {
+    const { outputFS } = options;
+
+    const styleRules = new Set<string>();
+
+    // Traverse the descendants of HTML bundle
+    // Extract the stylesRules from assets
+    bundleGraph.traverseBundles((childBundle) => {
+      childBundle.traverseAssets((asset) => {
+        const rules = asset.meta.styleRules;
+        if (rules == null) {
+          return;
+        }
+
+        assert(rules instanceof Array);
+
+        for (const rule of rules) {
+          styleRules.add(rule as string);
+        }
+      });
+    }, bundle);
+
+    if (styleRules.size === 0) return { contents, map };
+
+    const stylesheet = sort(Array.from(styleRules).join(''));
+
+    let newContents = '';
+
+    if (config.inlineCss) {
+      newContents = (
+        await posthtml()
+          .use(
+            insertAt({
+              selector: 'head',
+              append: '<style>' + stylesheet + '</style>',
+              behavior: 'inside',
+            })
+          )
+          .process(contents.toString())
+      ).html;
+    } else {
+      const { distDir } = bundle.target;
+
+      if (!outputFS.existsSync(distDir)) {
+        await outputFS.mkdirp(distDir);
+      }
+
+      const cssFileName = basename(bundle.displayName, '.html') + '.' + hash(stylesheet) + '.css';
+
+      await outputFS.writeFile(
+        join(distDir, cssFileName),
+        stylesheet,
+        undefined // for TypeScript
+      );
+
+      newContents = (
+        await posthtml()
+          .use(
+            insertAt({
+              selector: 'head',
+              append:
+                '<link href="' + bundle.target.publicUrl + cssFileName + '" rel="stylesheet" />',
```

Comment:
Why was this previously manually added and now is automated?

Reply:
Sorry, I don't get what you mean. Do you mean it was manually added in the fixture app https://github.com/atlassian-labs/compiled/pull/1303/files/83f34318edcce33f1d9ea0a95aa8f38a0cb70721#diff-8d93de8220a5ed7650b00d8791683374942b49b5fa6753bb833ba978af721596?

Reply:
Yep!

Reply:
Previously the fixture app didn't use `@compiled/react`, so for simplicity, it manually adds a CSS file in order to test that optimizer can sort the CSS. 
Now I cannot test optimizer independently because optimizer needs the metadata from transformer. I've added `@compiled/react` and `@compiled/parcel-transformer` to the fixture app. The test has become an integration test.
Worth mentioning [this file ](https://github.com/atlassian-labs/compiled/blob/master/fixtures/parcel-transformer-test-app/src/index.html) is the entrypoint (similar to webpack.config.js). Because previously the entrypoint references a CSS file, Parcel automatically adds CSS file to the output HTML.
Now since we've used CSS-in-JS (@compiled/react), we need to manually add the CSS file to output HTML.

I hope it makes sense 😄

Reply:
Yeahhh, I think I get it now. Thanks for the explanation, I was just curious :)